workflow-runtime/src/iosTest/kotlin/com/squareup/workflow1/internal/ThreadLocalTest.kt

```diff
@@ -0,0 +1,82 @@
+package com.squareup.workflow1.internal
+
+import platform.Foundation.NSLock
+import platform.Foundation.NSThread
+import kotlin.concurrent.Volatile
+import kotlin.test.Test
+import kotlin.test.assertEquals
+
+class ThreadLocalTest {
+
+  @Volatile
+  private var valueFromThread: Int = -1
+
+  @Test fun initialValue() {
+    val threadLocal = ThreadLocal(initialValue = { 42 })
+    assertEquals(42, threadLocal.get())
+  }
+
+  @Test fun settingValue() {
+    val threadLocal = ThreadLocal(initialValue = { 42 })
+    threadLocal.set(0)
+    assertEquals(0, threadLocal.get())
+  }
+
+  @Test fun initialValue_inSeparateThread_afterChanging() {
+    val threadLocal = ThreadLocal(initialValue = { 42 })
+    threadLocal.set(0)
+
+    val thread = NSThread {
+      valueFromThread = threadLocal.get()
+    }
+    thread.start()
+    thread.join()
+
+    assertEquals(42, valueFromThread)
+  }
+
+  @Test fun set_fromDifferentThreads_doNotConflict() {
+    val threadLocal = ThreadLocal(initialValue = { 0 })
+    val threadStartedLatch = NSLock().apply { lock() }
+    val firstReadLatch = NSLock().apply { lock() }
+    val firstReadDoneLatch = NSLock().apply { lock() }
+    val secondReadLatch = NSLock().apply { lock() }
+
+    val thread = NSThread {
+      threadStartedLatch.unlock()
```

Comment:
i'm a bit surprised this works without some assertion failure, as the lock API typically require that unlocks happen on the same thread that acquired the lock.

i'd like to offer some possible alternatives, but i'm having trouble figuring out where these 'platform' imports are defined, so am not sure what's available (e.g. semaphores might be the way to go here). anyone know where i might try looking?

Reply:
Should we pair on this?